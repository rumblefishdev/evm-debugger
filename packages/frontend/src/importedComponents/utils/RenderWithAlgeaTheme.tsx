import { ThemeProvider, CssBaseline } from '@mui/material'
import React from 'react'

import { theme } from '../theme/algaeTheme'

export const RenderWithAlgeaTheme: React.FC<React.PropsWithChildren> = ({
  children,
  ...props
}) => {
  return (
    <ThemeProvider theme={theme}>
      <CssBaseline />
      {children}
    </ThemeProvider>
  )
}
