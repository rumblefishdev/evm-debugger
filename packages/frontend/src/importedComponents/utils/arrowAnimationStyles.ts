const arrowHiddenMouseOn = {
  transform: 'translate(0px)',
}
const arrowVisibleMouseOff = {
  transition: 'none',
  transform: 'translate(0px)',
}
// eslint-disable-next-line import/exports-last
export type TDirection = 'left' | 'top' | 'right' | 'down'

const generateArrowHiddenMouseOff = (direction: TDirection, pixels: string) => {
  switch (direction) {
    case 'top': {
      return {
        transition: 'none',
        transform: `translateY(${pixels})`,
      }
    }
    case 'right': {
      return {
        transition: 'none',
        transform: `translateX(-${pixels})`,
      }
    }
    case 'down': {
      return {
        transition: 'none',
        transform: `translateY(-${pixels})`,
      }
    }
    case 'left': {
      return {
        transition: 'none',
        transform: `translateX(${pixels})`,
      }
    }
    default: {
      return { transition: 'none' }
    }
  }
}
const generateArrowVisibleMouseOn = (direction: TDirection, pixels: string) => {
  switch (direction) {
    case 'top': {
      return {
        transform: `translateY(-${pixels})`,
      }
    }
    case 'right': {
      return {
        transform: `translateX(${pixels})`,
      }
    }
    case 'down': {
      return {
        transform: `translateY(${pixels})`,
      }
    }
    case 'left': {
      return {
        transform: `translateX(-${pixels})`,
      }
    }
    default: {
      return { transform: 'translateX(0px)' }
    }
  }
}

export const generateArrowAnimation = (direction: TDirection, pixels = '50px') => {
  const arrowVisibleMouseOn = generateArrowVisibleMouseOn(direction, pixels)
  const arrowHiddenMouseOff = generateArrowHiddenMouseOff(direction, pixels)

  return {
    arrowVisibleMouseOn,
    arrowVisibleMouseOff,
    arrowHiddenMouseOn,
    arrowHiddenMouseOff,
  }
}
