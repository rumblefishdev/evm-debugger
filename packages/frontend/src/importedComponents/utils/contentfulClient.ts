import { ContentfulClient } from 'react-contentful'

export const contentfulClient = ContentfulClient({
  space: process.env.REACT_APP_CONTENTFUL_SPACE_ID,
  environment: process.env.REACT_APP_CONTENTFUL_ENVIRONMENT,
  accessToken: process.env.REACT_APP_CONTENTFUL_ACCESS_TOKEN,
})
