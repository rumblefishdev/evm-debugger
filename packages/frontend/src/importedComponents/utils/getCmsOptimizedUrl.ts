type CmsOptimized = {
  url: string
  quality?: number
  width?: number
  height?: number
}

// This will return url with additional tags, for images in lower quality and size
// https://www.contentful.com/developers/docs/references/images-api/#/reference/image-manipulation/quality
// https://www.contentful.com/developers/docs/references/images-api/#/reference/resizing-&-cropping/change-the-resizing-behavior

export const getCmsOptimizedUrl = ({
  url,
  quality = 60,
  width,
  height,
}: CmsOptimized): string => {
  let baseUrl = `${url}?fm=jpg&q=${quality}`

  if (width) baseUrl += `&w=${width.toString(10)}`

  if (height) baseUrl += `&h=${height.toString(10)}`

  if (width || height) baseUrl += '&fit=pad'

  return baseUrl
}
