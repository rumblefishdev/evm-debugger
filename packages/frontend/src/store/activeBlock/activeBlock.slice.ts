import type { PayloadAction } from '@reduxjs/toolkit'
import { createSlice } from '@reduxjs/toolkit'

import type { TMainTraceLogsWithId } from '../../types'
import type { TRootState } from '../store'
import { StoreKeys } from '../store.keys'

const initialState: null | TMainTraceLogsWithId = null

export const activeBlockSlice = createSlice({
  reducers: {
    loadActiveBlock: (state, action: PayloadAction<TMainTraceLogsWithId>) => {
      return action.payload
    },
  },
  name: StoreKeys.ACTIVE_BLOCK,
  initialState,
})

export const activeBlockReducer = activeBlockSlice.reducer

export const selectActiveBlock = (state: TRootState) => state.activeBlock

export const { loadActiveBlock } = activeBlockSlice.actions
