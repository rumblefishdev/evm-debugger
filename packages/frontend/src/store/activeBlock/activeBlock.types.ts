import type {
  TCallTypeOpcodes,
  TCreateTypeOpcodes,
  TStorageLogs,
} from '@evm-debuger/types'

import type { TParsedEventLog } from '../../types'

export type TParsedParams = {
  value: string
  type: string
  name: string
}

export type TBlockDefaultData = {
  value: string
  type: TCallTypeOpcodes | TCreateTypeOpcodes
  stackTrace: string
  passedGas: number
  gasCost: number
  blockNumber: string
  address: string
  isSuccess: boolean
}
export type TBlockCallSpecificData = {
  storageLogs: TStorageLogs
  storageAddress: string
  parsedOutput: null | TParsedParams[]
  parsedInput: null | TParsedParams[]
  parsedEvents: TParsedEventLog[]
  parsedError: null | TParsedParams[]
  isContract: boolean
  functionSignature: null | string
  errorSignature: null | string
  input: string
  output: string
}
export type TBlockCreateSpecificData = {
  storageLogs: TStorageLogs
  storageAddress: string
  input: string
  salt: string
}

export type TParsedActiveBlock = {
  defaultData: null | TBlockDefaultData
  createSpecificData: null | TBlockCreateSpecificData
  callSpecificData: null | TBlockCallSpecificData
}
