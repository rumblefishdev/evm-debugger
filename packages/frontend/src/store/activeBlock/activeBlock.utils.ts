// TODO: remove ts-ignore and fix typescript errors
// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-nocheck
import type { BytesLike, Result, ParamType } from 'ethers'
import { formatEther, isBytesLike } from 'ethers'

import { isArrayOfStrings } from '../../helpers/helpers'

export const safeArgParse = (arg: string | bigint | boolean | string[] | BytesLike | bigint[] | number) => {
  if (typeof arg === 'string') return arg

  if (BigInt(arg)) return `${formatEther(BigInt(arg).toString())} ETH`

  if (typeof arg === 'boolean') return arg ? 'true' : 'false'

  if (isBytesLike(arg)) return arg.toString()

  if (typeof arg === 'number') return arg

  if (isArrayOfStrings(arg)) return arg

  if (Array.isArray(arg) && arg.every((item) => BigInt(item))) return arg.map((item) => `${formatEther(BigInt(item).toString())} ETH`)
}

export const parseParameter = (parameterType, parameterValue) => {
  let parsedValues

  if (typeof parameterValue === 'string') parsedValues = safeArgParse(parameterValue)
  else if (parameterType.type === 'tuple[]')
    parsedValues = parameterValue.map((value) => parseParameter(parameterType.arrayChildren, value))
  else if (parameterType.baseType === 'array') parsedValues = parameterValue.map((value) => safeArgParse(value))
  else if (parameterType.baseType === 'tuple')
    parsedValues = parameterType.components.map((component, index) => {
      return parseParameter(component, parameterValue[index])
    })
  else parsedValues = safeArgParse(parameterValue)

  return {
    value: parsedValues,
    type: parameterType.type,
    name: parameterType.name,
  }
}

export const parseParameters = (params: ParamType[], result: Result) => {
  return params.map((parameterType, index) => {
    if (result) {
      const parameterValue = result[index]
      return parseParameter(parameterType, parameterValue)
    }

    return {
      value: 'Failed to decode',
      type: parameterType.type,
      name: parameterType.name,
    }
  })
}
