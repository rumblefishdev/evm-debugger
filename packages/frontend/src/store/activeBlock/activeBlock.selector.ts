import { checkIfOfCallType, checkIfOfCreateType } from '@evm-debuger/analyzer'
import type { TEventInfo } from '@evm-debuger/types'
import { createSelector } from '@reduxjs/toolkit'

import { getSignature, parseStackTrace } from '../../helpers/helpers'
import type { TParsedEventLog } from '../../types'
import type { TRootState } from '../store'
import type { TMainTraceLogsWithId } from '../traceLogs/traceLogs.types'
import { StoreKeys } from '../store.keys'
import { selectReducer } from '../store.utils'
import { contractNamesSelectors } from '../contractNames/contractNames.selectors'

import type { TBlockCallSpecificData, TParsedActiveBlock } from './activeBlock.types'
import { parseParameter, parseParameters } from './activeBlock.utils'

const selectActiveBlockState = createSelector([selectReducer(StoreKeys.ACTIVE_BLOCK)], (state) => state)

const parseEventLog = (eventLogs: TEventInfo[]): TParsedEventLog[] => {
  return eventLogs.map((eventLog) => {
    if (!eventLog.eventDescription) return { signature: null, parsedArgs: null, name: null }
    const { eventDescription } = eventLog
    const { name, signature, args, eventFragment } = eventDescription
    const { inputs } = eventFragment
    const parsedArgs = inputs.map((parameterType, index) => {
      const parameterValue = args[index]
      return parseParameter(parameterType, parameterValue)
    })

    return { signature, parsedArgs, name }
  })
}

const parseActiveBlock = (block: TMainTraceLogsWithId, contractName: string | null) => {
  const result: TParsedActiveBlock = {
    defaultData: null,
    createSpecificData: null,
    callSpecificData: null,
  }

  const { address, gasCost, passedGas, stackTrace, type, value, blockNumber, isSuccess, startIndex, returnIndex } = block
  result.defaultData = {
    value,
    type,
    startIndex,
    stackTrace: parseStackTrace(stackTrace),
    returnIndex,
    passedGas,
    isSuccess,
    gasCost,
    blockNumber,
    address,
  }

  // eslint-disable-next-line unicorn/consistent-destructuring
  if (checkIfOfCallType(block) && block.isContract) {
    const {
      events,
      functionFragment,
      errorDescription,
      decodedInput,
      decodedOutput,
      storageAddress,
      storageLogs,
      input,
      output,
      isContract,
    } = block

    const callResult: TBlockCallSpecificData = {
      storageLogs,
      storageAddress,
      parsedOutput: null,
      parsedInput: null,
      parsedEvents: parseEventLog(events),
      parsedError: null,
      output,
      isContract,
      input,
      functionSignature: null,
      errorSignature: null,
      contractName,
    }

    if (functionFragment) {
      const { inputs, outputs } = functionFragment

      const parsedInput = parseParameters(inputs, decodedInput)
      const parsedOutput = parseParameters(outputs, decodedOutput)

      const signature = getSignature(functionFragment)

      callResult.parsedInput = parsedInput
      callResult.parsedOutput = parsedOutput
      callResult.functionSignature = signature
    }

    if (errorDescription) {
      const { signature, errorFragment, args } = errorDescription
      const { inputs } = errorFragment

      const parsedError = parseParameters(inputs, args)

      callResult.errorSignature = signature
      callResult.parsedError = parsedError
    }

    result.callSpecificData = callResult
  }

  if (checkIfOfCreateType(block)) {
    const { storageAddress, storageLogs, salt, input } = block

    result.createSpecificData = { storageLogs, storageAddress, salt, input }
  }

  return result
}

export const selectParsedActiveBlock = createSelector(
  [selectActiveBlockState, contractNamesSelectors.selectAll],
  (activeBlock, contractNames) => {
    const contract = contractNames.find((contractName) => contractName.address === activeBlock.address)
    const contractNameString = contract?.contractName ?? null
    return parseActiveBlock(activeBlock, contractNameString)
  },
)

export const selectActiveBlock = createSelector([selectActiveBlockState], (state) => state)

export const getTraceLogErrorOutput = (block: TMainTraceLogsWithId) => {
  const errorSignature = checkIfOfCallType(block) ? block.errorDescription?.signature : null

  return errorSignature ? errorSignature : 'Revert (no revert message was provided)'
}

export const activeBlockSelectors = { selectParsedActiveBlock, selectActiveBlock }
