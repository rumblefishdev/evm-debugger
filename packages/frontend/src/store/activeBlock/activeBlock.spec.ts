import { BigNumber, ethers } from 'ethers'

import { parseParameter, parseParameters } from './activeBlock.utils'

const ADDRESS_TYPE = ethers.utils.ParamType.fromObject({
  type: 'address',
  name: 'recipient',
  indexed: false,
  components: null,
  baseType: 'address',
  arrayLength: null,
  arrayChildren: null,
  _isParamType: true,
})

const UINT256_TYPE = ethers.utils.ParamType.fromObject({
  type: 'uint256',
  name: null,
  indexed: null,
  components: null,
  baseType: 'uint256',
  arrayLength: null,
  arrayChildren: null,
  _isParamType: true,
})

const BYTES_TYPE = ethers.utils.ParamType.fromObject({
  type: 'bytes',
  name: 'permit',
  indexed: null,
  components: null,
  baseType: 'bytes',
  arrayLength: null,
  arrayChildren: null,
  _isParamType: true,
})

const ADDRESS_ARRAY_TYPE = ethers.utils.ParamType.fromObject({
  type: 'address[]',
  name: 'srcReceivers',
  indexed: null,
  components: null,
  baseType: 'array',
  arrayLength: -1,
  arrayChildren: ADDRESS_TYPE,
  _isParamType: true,
})

const UNIT256_ARRAY_TYPE = ethers.utils.ParamType.fromObject({
  type: 'uint256[]',
  name: 'srcAmounts',
  indexed: null,
  components: null,
  baseType: 'array',
  arrayLength: -1,
  arrayChildren: UINT256_TYPE,
  _isParamType: true,
})

const TUPLE_TYPE = ethers.utils.ParamType.fromObject({
  type: 'tuple',
  name: 'desc',
  indexed: null,
  components: [
    ADDRESS_TYPE,
    UINT256_TYPE,
    BYTES_TYPE,
    ADDRESS_ARRAY_TYPE,
    UNIT256_ARRAY_TYPE,
  ],
  baseType: 'tuple',
  arrayLength: null,
  arrayChildren: null,
  _isParamType: true,
})

const ARRAY_TYPE = {
  type: 'tuple[]',
  name: 'consideration',
  indexed: false,
  components: [
    ADDRESS_TYPE,
    UINT256_TYPE,
    BYTES_TYPE,
    ADDRESS_ARRAY_TYPE,
    UNIT256_ARRAY_TYPE,
  ],
  baseType: 'array',
  arrayLength: -1,
  arrayChildren: TUPLE_TYPE,
  _isParamType: true,
}

describe('parseParameters tests', () => {
  describe('Exceptions', () => {
    it('Address exist but not with failed to decode output', () => {
      const result = parseParameters([ADDRESS_TYPE], null)
      expect(result).toStrictEqual([
        {
          value: 'Failed to decode',
          type: 'address',
          name: 'recipient',
        },
      ])
    })

    it('Tuple parameter exist but not with failed to decode output', () => {
      const result = parseParameters([TUPLE_TYPE], null)
      expect(result).toStrictEqual([
        {
          value: 'Failed to decode',
          type: 'tuple',
          name: 'desc',
        },
      ])
    })
  })
})

describe('parseParameter tests', () => {
  describe('Basic types', () => {
    it('Address type', () => {
      const value = '0xd45C07b37538e94cf0dC768A56491EeC14Ce898B'
      const result = parseParameter(
        ADDRESS_TYPE,
        '0xd45C07b37538e94cf0dC768A56491EeC14Ce898B',
      )
      expect(result).toStrictEqual({
        value,
        type: 'address',
        name: 'recipient',
      })
    })

    it('unit256 type', () => {
      const value = BigNumber.from('1')
      const result = parseParameter(UINT256_TYPE, value)
      expect(result).toStrictEqual({
        value: '0.000000000000000001 ETH',
        type: 'uint256',
        name: null,
      })
    })

    it('Bytes type', () => {
      const value =
        '0x00000000000000000000000096c195f6643a3d797cb90cb6ba0ae2776d51b5f30000000000000000000000000000000000000000000000000000000000000001000000000000000000000000000000000000000000000000000000000000001e'
      const result = parseParameter(BYTES_TYPE, value)
      expect(result).toStrictEqual({
        value,
        type: 'bytes',
        name: 'permit',
      })
    })
  })
  describe('Collection types', () => {
    it('Address array', () => {
      const address1 = '0xd45C07b37538e94cf0dC768A56491EeC14Ce898B'
      const address2 = '0xabcd07b37538e94cf0dC768A56491EeC14Ce898B'
      const values = [address1, address2]
      const result = parseParameter(ADDRESS_ARRAY_TYPE, values)
      expect(result).toStrictEqual({
        value: values,
        type: 'address[]',
        name: 'srcReceivers',
      })
    })

    it('uint256 array', () => {
      const values = [BigNumber.from('1'), BigNumber.from('2')]
      const result = parseParameter(UNIT256_ARRAY_TYPE, values)
      expect(result).toStrictEqual({
        value: ['0.000000000000000001 ETH', '0.000000000000000002 ETH'],
        type: 'uint256[]',
        name: 'srcAmounts',
      })
    })

    it('Tuple type', () => {
      const bytes =
        '0x00000000000000000000000096c195f6643a3d797cb90cb6ba0ae2776d51b5f30000000000000000000000000000000000000000000000000000000000000001000000000000000000000000000000000000000000000000000000000000001e'
      const values = [
        '0x30dcBa0405004cF124045793E1933C798Af9E66a',
        BigNumber.from('1'),
        bytes,
        [
          '0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE',
          '0xEaaaaaEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE',
        ],
        [BigNumber.from('2'), BigNumber.from('3')],
      ]
      const result = parseParameter(TUPLE_TYPE, values)
      expect(result).toMatchSnapshot()
    })
    it('Tuple[] type', () => {
      const bytes =
        '0x00000000000000000000000096c195f6643a3d797cb90cb6ba0ae2776d51b5f30000000000000000000000000000000000000000000000000000000000000001000000000000000000000000000000000000000000000000000000000000001e'

      const values = [
        [
          '0x30dcBa0405004cF124045793E1933C798Af9E66a',
          BigNumber.from('1'),
          bytes,
          [
            '0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE',
            '0xEaaaaaEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE',
          ],
          [BigNumber.from('2'), BigNumber.from('3')],
        ],
        [
          '0xbbbBa0405004cF124045793E1933C798Af9E66a',
          BigNumber.from('5'),
          bytes,
          [
            '0xEeeeeEeeeEeEeeEeEeEeeffffffffffFAFFFfffE',
            '0xEaaaaaEeEeeEeEeEeeEffffffffffffffffffff',
          ],
          [BigNumber.from('62'), BigNumber.from('33')],
        ],
      ]
      const result = parseParameter(ARRAY_TYPE, values)
      expect(result).toMatchSnapshot()
    })
  })
})
