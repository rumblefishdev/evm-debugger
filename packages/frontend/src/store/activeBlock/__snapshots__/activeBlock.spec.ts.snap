// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseParameter tests Collection types Tuple type 1`] = `
Object {
  "name": "desc",
  "type": "tuple",
  "value": Array [
    Object {
      "name": "recipient",
      "type": "address",
      "value": "0x30dcBa0405004cF124045793E1933C798Af9E66a",
    },
    Object {
      "name": "",
      "type": "uint256",
      "value": "0.000000000000000001 ETH",
    },
    Object {
      "name": "permit",
      "type": "bytes",
      "value": "0x00000000000000000000000096c195f6643a3d797cb90cb6ba0ae2776d51b5f30000000000000000000000000000000000000000000000000000000000000001000000000000000000000000000000000000000000000000000000000000001e",
    },
    Object {
      "name": "srcReceivers",
      "type": "address[]",
      "value": Array [
        "0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE",
        "0xEaaaaaEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE",
      ],
    },
    Object {
      "name": "srcAmounts",
      "type": "uint256[]",
      "value": Array [
        "0.000000000000000002 ETH",
        "0.000000000000000003 ETH",
      ],
    },
  ],
}
`;

exports[`parseParameter tests Collection types Tuple[] type 1`] = `
Object {
  "name": "consideration",
  "type": "tuple[]",
  "value": Array [
    Object {
      "name": "desc",
      "type": "tuple",
      "value": Array [
        Object {
          "name": "recipient",
          "type": "address",
          "value": "0x30dcBa0405004cF124045793E1933C798Af9E66a",
        },
        Object {
          "name": "",
          "type": "uint256",
          "value": "0.000000000000000001 ETH",
        },
        Object {
          "name": "permit",
          "type": "bytes",
          "value": "0x00000000000000000000000096c195f6643a3d797cb90cb6ba0ae2776d51b5f30000000000000000000000000000000000000000000000000000000000000001000000000000000000000000000000000000000000000000000000000000001e",
        },
        Object {
          "name": "srcReceivers",
          "type": "address[]",
          "value": Array [
            "0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE",
            "0xEaaaaaEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE",
          ],
        },
        Object {
          "name": "srcAmounts",
          "type": "uint256[]",
          "value": Array [
            "0.000000000000000002 ETH",
            "0.000000000000000003 ETH",
          ],
        },
      ],
    },
    Object {
      "name": "desc",
      "type": "tuple",
      "value": Array [
        Object {
          "name": "recipient",
          "type": "address",
          "value": "0xbbbBa0405004cF124045793E1933C798Af9E66a",
        },
        Object {
          "name": "",
          "type": "uint256",
          "value": "0.000000000000000005 ETH",
        },
        Object {
          "name": "permit",
          "type": "bytes",
          "value": "0x00000000000000000000000096c195f6643a3d797cb90cb6ba0ae2776d51b5f30000000000000000000000000000000000000000000000000000000000000001000000000000000000000000000000000000000000000000000000000000001e",
        },
        Object {
          "name": "srcReceivers",
          "type": "address[]",
          "value": Array [
            "0xEeeeeEeeeEeEeeEeEeEeeffffffffffFAFFFfffE",
            "0xEaaaaaEeEeeEeEeEeeEffffffffffffffffffff",
          ],
        },
        Object {
          "name": "srcAmounts",
          "type": "uint256[]",
          "value": Array [
            "0.000000000000000062 ETH",
            "0.000000000000000033 ETH",
          ],
        },
      ],
    },
  ],
}
`;
